### PR TITLE
Taoshaf.check se2

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by fantaosha** [comment]
Now `autograd` works for `SE2.log()` and `SE2.exp()` using `torch.where()`.

File: theseus/geometry/se2.py [hunk 1]
```diff
@@ -92,37 +94,25 @@ def update_from_rot_and_trans(self, rotation: SO2, translation: Point2):
...
-        halftheta_by_tan_of_halftheta[idx_regular_vals] = (
-            -(half_theta * sine)[idx_regular_vals] / cos_minus_one[idx_regular_vals]
+        # Compute the approximations when theta is near to 0
+        small_theta = theta.abs() < SE2.SE2_EPS
+        non_zero = torch.ones(1, dtype=self.dtype, device=self.device)
+        sine_nz = torch.where(small_theta, non_zero, sine)
+        a = (
+            0.5
+            * (1 + cosine)
+            * torch.where(small_theta, 1 + sine ** 2 / 6, theta / sine_nz)
```
Comment (fantaosha) [inline theseus/geometry/se2.py:107]: `theta / sine_nz` instead of `theta / sine` is critical to avoid `NaN` errors.

File: theseus/geometry/se2.py [hunk 2]
```diff
@@ -133,32 +123,21 @@ def exp_map(tangent_vector: torch.Tensor) -> LieGroup:
...
-                0.5 * small_theta - small_theta / 24 * small_theta_sq
-            )
+        # Compute the approximations when theta is near to 0
+        small_theta = theta.abs() < SE2.SE2_EPS
+        non_zero = torch.ones(
+            1, dtype=tangent_vector.dtype, device=tangent_vector.device
+        )
+        theta_nz = torch.where(small_theta, non_zero, theta)
+        a = torch.where(
+            small_theta, -theta / 2 + theta ** 3 / 24, (cosine - 1) / theta_nz
```
Comment (fantaosha) [inline theseus/geometry/se2.py:133]: Similar to my previous [comment](https://github.com/facebookresearch/theseus/pull/49/files#r794204182), use `(cosine - 1) / theta_nz` rather than `(cosine - 1) / theta` to avoid `NaN` errors for `autograd`. @luisenp

File: theseus/geometry/se2.py [hunk 3]
```diff
@@ -92,37 +94,25 @@ def update_from_rot_and_trans(self, rotation: SO2, translation: Point2):
     # From https://github.com/strasdat/Sophus/blob/master/sophus/se2.hpp#L160
```
Comment (luisenp) [inline theseus/geometry/se2.py:94]: How does the approximation you used below relate to the one used in Sophus? Can you change the reference to point to the source for the expressions below?

File: theseus/geometry/se2.py [hunk 4]
```diff
@@ -92,37 +94,25 @@ def update_from_rot_and_trans(self, rotation: SO2, translation: Point2):
...
 
-        # Compute halftheta_by_tan_of_halftheta when theta is not near zero
-        idx_regular_vals = cos_minus_one.abs() > theseus.constants.EPS
-        halftheta_by_tan_of_halftheta[idx_regular_vals] = (
-            -(half_theta * sine)[idx_regular_vals] / cos_minus_one[idx_regular_vals]
+        # Compute the approximations when theta is near to 0
+        small_theta = theta.abs() < SE2.SE2_EPS
+        non_zero = torch.ones(1, dtype=self.dtype, device=self.device)
+        sine_nz = torch.where(small_theta, non_zero, sine)
+        a = (
```
Comment (luisenp) [inline theseus/geometry/se2.py:104]: nit: a --> `theta_by_sine`

File: theseus/geometry/se2.py [hunk 5]
```diff
@@ -133,32 +123,21 @@ def exp_map(tangent_vector: torch.Tensor) -> LieGroup:
...
-            one_minus_cos_theta_by_theta[idx_small_thetas] = (
-                0.5 * small_theta - small_theta / 24 * small_theta_sq
-            )
+        # Compute the approximations when theta is near to 0
+        small_theta = theta.abs() < SE2.SE2_EPS
+        non_zero = torch.ones(
+            1, dtype=tangent_vector.dtype, device=tangent_vector.device
+        )
+        theta_nz = torch.where(small_theta, non_zero, theta)
+        a = torch.where(
```
Comment (luisenp) [inline theseus/geometry/se2.py:132]: Maybe add a quick comment that `a ~ (cos - 1) / theta` so that it's easier to read. I prefer this to the `cos_minus_one_by_theta` style that we had before. Similar comment for `b`.

File: theseus/geometry/se2.py [hunk 6]
```diff
@@ -92,37 +94,25 @@ def update_from_rot_and_trans(self, rotation: SO2, translation: Point2):
...
-        v_inv[:, 1, 1] = halftheta_by_tan_of_halftheta
-        tangent_translation = torch.matmul(v_inv, self[:, :2].unsqueeze(-1))
+        # Compute the translation
+        ux = a * self[:, 0] + b * self[:, 1]
+        uy = a * self[:, 1] - b * self[:, 0]
 
-        return torch.cat([tangent_translation.view(-1, 2), theta], dim=1)
+        return torch.stack((ux, uy, theta), dim=1)
 
     # From https://github.com/strasdat/Sophus/blob/master/sophus/se2.hpp#L558
```
Comment (luisenp) [inline theseus/geometry/se2.py:117]: Similar comment about the source.

**Review by luisenp — COMMENTED** [review]
LGTM, thanks for the fix! Did you test this in the tactile pushing example?

File: theseus/geometry/se2.py (hunk 3 above)
Comment (fantaosha) [reply theseus/geometry/se2.py:94]: I forgot to remove this line. The implementation of `so2.exp_map` and `so2.log_map` is a rewrite. I will add some comments.

File: theseus/geometry/se2.py (hunk 5 above)
Comment (fantaosha) [reply theseus/geometry/se2.py:132]: Will do. Thanks for the reminder.

File: theseus/geometry/se2.py (hunk 3 above)
Comment (fantaosha) [reply theseus/geometry/se2.py:94]: Yes, the tactile pushing example successfully computes the gradient.
